components/director/internal/tenantfetchersvc/provisioner.go

```diff
@@ -11,78 +11,112 @@ import (
+	if len(request.SubaccountTenantID) > 0 {
+		return fmt.Errorf("tenant with ID %s is of type %s and supports only regional provisioning", request.SubaccountTenantID, tenantEntity.Subaccount)
 	}
 
-	if err := p.tenantSvc.CreateManyIfNotExists(ctx, tenant); err != nil {
-		if !apperrors.IsNotUniqueError(err) {
-			return errors.Wrapf(err, tenantCreationFailureMsgFmt, externalTenantID)
-		}
+	tenants, err := p.tenantsFromRequest(ctx, request)
+	if err != nil {
+		return err
 	}
 
-	return nil
+	return p.tenantSvc.CreateManyIfNotExists(ctx, tenants...)
 }
 
-func (p *provisioner) ensureParentExists(ctx context.Context, parentTenantID, childTenantID string) (string, error) {
-	log.C(ctx).Infof("Ensuring parent tenant with external ID %s for tenant with external ID %s exists", parentTenantID, childTenantID)
-	id, err := p.tenantSvc.GetInternalTenant(ctx, parentTenantID)
-	if err != nil && !apperrors.IsNotFoundError(err) && err != sql.ErrNoRows {
-		return "", errors.Wrapf(err, "failed to retrieve internal ID of parent with external ID %s", parentTenantID)
-	}
-	if id != "" {
-		log.C(ctx).Infof("Parent tenant with external ID %s already exists", parentTenantID)
-		return id, nil
+func (p *provisioner) ProvisionRegionalTenants(ctx context.Context, request TenantProvisioningRequest) error {
+	tenants, err := p.tenantsFromRequest(ctx, request)
+	if err != nil {
+		return err
 	}
+	return p.tenantSvc.CreateManyIfNotExists(ctx, tenants...)
+}
+
+func (p *provisioner) tenantsFromRequest(ctx context.Context, request TenantProvisioningRequest) ([]model.BusinessTenantMappingInput, error) {
+	tenants := make([]model.BusinessTenantMappingInput, 0)
+
+	accountTenant := p.newAccountTenant(request.AccountTenantID, request.CustomerTenantID, request.Subdomain, request.Region)
 
-	log.C(ctx).Infof("Creating parent tenant with external ID %s", parentTenantID)
-	err = p.tenantSvc.CreateManyIfNotExists(ctx, p.customerTenant(parentTenantID))
-	if err != nil && apperrors.IsNotUniqueError(err) {
-		log.C(ctx).Infof("Parent tenant with external ID %s already exists", parentTenantID)
-		return p.tenantSvc.GetInternalTenant(ctx, parentTenantID)
-	} else if err != nil {
-		return "", errors.Wrapf(err, "failed to create parent tenant with ID %s", parentTenantID)
+	if len(request.SubaccountTenantID) > 0 {
+		intTenantID, err := p.tenantSvc.GetInternalTenant(ctx, request.AccountTenantID)
+		if err != nil && !apperrors.IsNotFoundError(err) {
+			return nil, errors.Wrapf(err, "while checking if parent of type account already exists")
+		}
+		subaccountTenant := p.newSubaccountTenant(request.SubaccountTenantID, request.AccountTenantID, request.Subdomain, request.Region)
+		if len(intTenantID) > 0 {
+			log.C(ctx).Infof("Parent tenant with internal ID %s is already present", intTenantID)
+			subaccountTenant.Parent = intTenantID
+			return []model.BusinessTenantMappingInput{subaccountTenant}, nil
+		}
+
+		accountTenant.Subdomain = ""
+		accountTenant.Region = ""
```

Comment:
Or it could get the region and subdomain from the SA, but I tought that later way we won't be sure if the labels are real, or are "inherited" from the SA